### ISSUE TITLE
Vaadin 8 - Binder "refresh" API

### ISSUE THREAD

**Issue by lorenzobenvenuti** [vaadin/framework#8097]
Hi,

I've read a blog post where you're encouraging people to try Vaadin 8 and give you feedback... so here I am! I'm playing with Vaadin 8 beta1 in order to understand if/when we can migrate our application from Vaadin 7.7 to 8. It is my understanding that most of the efforts have been aimed at rewriting data model and binding, so I started from there trying to write a simple master/detail ui (see attached sources). 
The first difficulty I encountered, coming from Vaadin 7, is keeping data aligned between model, `Grid` and detail. In Vaadin 7 when a `Property` was edited a `ValueChangeEvent`was fired, resulting in the `Grid` cell (or `Table` cell, or other `Field` bound to the same `Property`)  refreshing automatically. At the same time we could make programmatic changes to the data provider and see the UI update accordingly. In Vaadin 8 I've been able to update the `Grid` when the detail changes (using a `StatusChangeListener` and refreshing the `Grid` with `grid.getDataCommunicator().refresh`), but I didn't find a clean way to update the detail when the data model has been changed programmatically. Let's suppose I want to blank a field depending on the value of another (in the example UI, blanking the awards list if the user unchecks the "Awards" field):

```java
binder.addStatusChangeListener(event -> {
    if (!binder.getBean().isAwarded()) {
        // changing the model...
        binder.getBean().setAwards(Collections.emptyList());
        // ... how to refresh the view!?
    }
});
```

I tried to use `binder.readBean(binder.getBean())`, but invoking `readBean` from a `StatusChangeListener` causes a loop; the only way I found to update the detail was to alter the `Field` value, and in my opinion this approach is suboptimal because in this way application logic has to deal with conversion/null representation. For the previous example:

```java
binder.addStatusChangeListener(event -> {
    if (!binder.getBean().isAwarded()) {
        awardsListField.setValue(""); // we have to deal with "", which is in the "presentation" domain!
    }
});
```

In other words: I'd like to change data, not UI components that are representing those data in a fancy way.

I ended up adding a `Binder.refresh` method that updates all the bound `Field` instances reading values from the underlying bean; it doesn't fire events (to prevent the loop described before when used inside a `StatusChangeListener`). If there's a better way to achieve the same goal please ignore this issue and the corresponding PR! 

Thank you,

lorenzo

[ui.zip](https://github.com/vaadin/framework/files/674378/ui.zip)

**Comment by lorenzobenvenuti** [comment]
Created PR: https://github.com/vaadin/framework/pull/8098

**Comment by elmot** [comment]
Hello, Lorenzo,

Thanks for your feedback. 
First of all, please sign our CLA (yellow button at PR page) to be able to contribute to the framework.
Actually I think exposing that internal update code (updating bindings without status change events) is not the best solution, misuse of that method can break an application logic.

And even in your case - silent change of any value without additional ChangeEvent round-trip might cause inconsistent validation status.

I see two possible ways to implement your logic.
1) Implement your synchronous change at DataProvider/Setter level. You can make a setter for `awarded` which also modifies `awards`, then the only one event round-trip will happen.
2) Prevent Binder to generate an infinite loop -

```
if (!binder.getBean().isAwarded() && !binder.getBean().getAwards().isEmpty()) {
        // changing the model...
        binder.getBean().setAwards(Collections.emptyList());
        binder.readBean(binder.getBean());
    }
```

**Comment by lorenzobenvenuti** [comment]
> First of all, please sign our CLA (yellow button at PR page) to be able to contribute to the framework.

Actually I've already signed it, I guess Github is having some issues because I pushed changes using a non-primary email account.

> Actually I think exposing that internal update code (updating bindings without status change events) is not the best solution, misuse of that method can break an application logic.

My idea was to give a way to refresh the view after the model has changed. Something in line with the `grid.getDataCommunicator().refresh` I had to invoke to make the `Grid` refresh after editing the detail.

> Implement your synchronous change at DataProvider/Setter level. You can make a setter for awarded which also modifies awards, then the only one event round-trip will happen.

Maybe I'm missing something, but this seems just a different way to edit my data model and not to update the UI with the changes (it would be just a refactor of the code of my first post).

> Prevent Binder to generate an infinite loop -

This workaround works, but it seems a bit verbose: in our application we have lots of "chained" fields: having to protect every field update with an `if` to avoid loops seems a bit cumbersome.

Thank you,

lorenzo

**Comment by Legioth** [comment]
I'm sharing the concern with things getting out of sync if `refresh` doesn't fire the appropriate events.

Another way of approaching this problem could be to use an event that fires only when something has actually changed. This could be achieved by changing the conditions for firing status change events so that they are only fired if something has actually changed, or alternatively use value change events instead since those should already be defined to only fire for genuine changes. Using value change events for this would be facilitated by a shorthand in `Binder` that adds the same listener to all bound fields. There are also other reasons for why this feature would be useful.

**Comment by lorenzobenvenuti** [comment]
So, if I understood correctly: 

* `Binder` should lister for `ValueChangeEvent` for all the fields and fire a `StatusChangeEvent`
* Inside a `StatusChangeListener` you could change the model state and invoke a `readBean`
* `readBean` will result on a new `StatusChangeListener` only if the content of a `Field` is changed, preventing loops

Is that right? I'm wondering if the `StatusChangeEvent` should be split in two different events, one to notify changes in validation state and the other for value changes.

**Comment by Legioth** [comment]
> * `Binder` should lister for `ValueChangeEvent` for all the fields and fire a `StatusChangeEvent`

I don't think so. `StateChangeEvent` should actually only be fired if the value changes in such as way that e.g. the validation status of that field changes.

The intention of `StatusChangeEvent` is that it should let you know when the _status_ of a `Binder` changes. The most common cases of this would be when the validation status of any of the bound field changes or when the return value of `hasChanges()` changes. The current implementation is however overly eager since it interprets almost any change as a potential status change and fires an event just to be on the safe side. This means that the event is currently also handy for finding out about other changes, although this is not the intention.

Many cases that currently use `StatusChangeEvent` should probably instead listen for `ValueChangeEvent`. It's not done in this way today since there's no convenient way of adding a value change listener to all bound fields. Before we can change `StatusChangeEvent` to only fire when some status actually changes, we should first add a shorthand for listening for `ValueChangeEvent` and update all relevant examples and demos to use that event instead, so that those use cases would still be supported.

> * Inside a `StatusChangeListener` you could change the model state and invoke a `readBean`
> * `readBean` will result on a new `StatusChangeListener` only if the content of a `Field` is changed, preventing loops

There would be no loops, but `StatusChangeEvent` would still not work for the case you described since there might be value changes that don't cause any `StatusChangeEvent` (e.g. if both the old and the new values pass validation). You case should instead use the proposed shorthand for listening for `ValueChangeEvent`. There would not be any loops in that case either since there would only be a `ValueChangeEvent` if `readBean` actually causes some `Field` value to change.

**Comment by pleku** [comment]
> You case should instead use the proposed shorthand for listening for ValueChangeEvent
vaadin/framework#39 is currently targeted for 8.1 but we might need to take a look at bringing it to 8.0 ?

**Comment by Legioth** [comment]
Targeting for 8.0 seems like the only reasonable approach to me, since we wouldn't want to reduce the scope of  `StatusChangeEvent` between 8.0 and 8.1.

Implementing vaadin/framework8-issues#39 would hopefully not require too much effort, but changing `StatusChangeEvent` to not fire unless something actually changes might require some additional tweaking.

**Comment by pleku** [comment]
Hi Lorenzo! Thanks for the feedback, we have considered this and are going to add vaadin/framework8-issues#39 which we think will help with the use case, and avoid trying to "misuse" the `StatusChangeHandler`.

Closing this ticket for now.

**Comment by lorenzobenvenuti** [comment]
Ok, thanks!
